crates/bgp-pkt/src/nlri/bgp_ls.rs

```diff
@@ -0,0 +1,753 @@
+    /// |                  Link Remote Identifier                       |
+    /// +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+    /// ```
+    /// see [RFC7752 Section 1.1](https://www.rfc-editor.org/rfc/rfc5307#section-1.1)
+    LinkLocalRemoteIdentifiers {
+        link_local_identifier: u32,
+        link_remote_identifier: u32,
+    },
+
+    IPv4InterfaceAddress(Ipv4Addr),
+    IPv4NeighborAddress(Ipv4Addr),
+
+    /// MUST NOT be local-link
+    IPv6InterfaceAddress(Ipv6Addr),
+
+    /// MUST NOT be local-link
+    IPv6NeighborAddress(Ipv6Addr),
+    MultiTopologyIdentifier(MultiTopologyIdData),
+    Unknown {
+        code: u16,
+        value: Vec<u8>,
+    },
+}
+
+impl BgpLsLinkDescriptorTlv {
+    pub const fn code(&self) -> Result<BgpLsLinkDescriptorTlvType, u16> {
+        match self {
+            BgpLsLinkDescriptorTlv::LinkLocalRemoteIdentifiers { .. } => {
+                Ok(BgpLsLinkDescriptorTlvType::LinkLocalRemoteIdentifiers)
+            }
+            BgpLsLinkDescriptorTlv::IPv4InterfaceAddress(..) => {
+                Ok(BgpLsLinkDescriptorTlvType::IPv4InterfaceAddress)
+            }
+            BgpLsLinkDescriptorTlv::IPv4NeighborAddress(..) => {
+                Ok(BgpLsLinkDescriptorTlvType::IPv4NeighborAddress)
+            }
+            BgpLsLinkDescriptorTlv::IPv6InterfaceAddress(..) => {
+                Ok(BgpLsLinkDescriptorTlvType::IPv6InterfaceAddress)
+            }
+            BgpLsLinkDescriptorTlv::IPv6NeighborAddress(..) => {
+                Ok(BgpLsLinkDescriptorTlvType::IPv6NeighborAddress)
+            }
+            BgpLsLinkDescriptorTlv::MultiTopologyIdentifier(..) => {
+                Ok(BgpLsLinkDescriptorTlvType::MultiTopologyIdentifier)
+            }
+            BgpLsLinkDescriptorTlv::Unknown { code, .. } => Err(*code),
+        }
+    }
+
+    pub const fn raw_code(&self) -> u16 {
+        match self.code() {
+            Ok(value) => value as u16,
+            Err(value) => value,
+        }
+    }
+}
+
+/// ```text
+/// +--------------------+-------------------+----------+
+/// | Sub-TLV Code Point | Description       |   Length |
+/// +--------------------+-------------------+----------+
+/// |        512         | Autonomous System |        4 |
+/// |        513         | BGP-LS Identifier |        4 |
+/// |        514         | OSPF Area-ID      |        4 |
+/// |        515         | IGP Router-ID     | Variable |
+/// |        516         | BGP Router-ID     |        4 |
+/// |        517         | Member-AS Number  |        4 |
+/// +--------------------+-------------------+----------+
+/// ```
+/// see [RFC7752 Section 3.2.1](https://www.rfc-editor.org/rfc/rfc7752#section-3.2.1)
+#[derive(Debug, Clone, Eq, PartialEq, Serialize, Deserialize)]
+#[cfg_attr(feature = "fuzz", derive(arbitrary::Arbitrary))]
+pub enum BgpLsNodeDescriptorSubTlv {
+    AutonomousSystem(u32),
+    BgpLsIdentifier(u32),
+    OspfAreaId(u32),
+    IgpRouterId(Vec<u8>),
```

Comment:
According to RFC 7752: `this contains a 6-octet ISO Node-ID` .. type should be `[u8; 6]`. I think in the tests you're using 7 elements vectors.

Reply:
well actually...

For an IS-IS non-pseudonode, this contains a <ins>6-octet</ins> ISO Node-ID (ISO system-
ID).  For an IS-IS pseudonode corresponding to a LAN, this
contains the 6-octet ISO Node-ID of the Designated Intermediate
System (DIS) followed by a 1-octet, nonzero PSN identifier (<ins>7
octets in total</ins>).  For an OSPFv2 or OSPFv3 non-pseudonode, this
contains the <ins>4-octet</ins> Router-ID.  For an OSPFv2 pseudonode
representing a LAN, this contains the 4-octet Router-ID of the
Designated Router (DR) followed by the 4-octet IPv4 address of the
DR's interface to the LAN (<ins>8 octets in total</ins>).  Similarly, for an
OSPFv3 pseudonode, this contains the 4-octet Router-ID of the DR
followed by the 4-octet interface identifier of the DR's interface
to the LAN (<ins>8 octets in total</ins>).  The TLV size in combination with
the protocol identifier enables the decoder to determine the type
of the node.

so *currently* it can range anywhere from 4 to 8 bytes but not 7... i figured a Vec would do the trick but i can make this a `[u8; 8]` if you prefer

Reply:
Ah, I've not read the full thing. Then it makes sense.